### PR TITLE
#134249297/Game group chat feature

### PR DESCRIPTION
**Pull request by andela-aolaniran**
#### What does this PR do?
Implementation of group chat feature
#### Description of Task to be completed?
Allow users in a game to chat with one another
#### How should this be manually tested?
N/A
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
Users should be able to chat with the group while they are in a game.
As a registered user in a game I want to be able to chat with the other players so that I can communicate with other players and have a more enjoyable experience using the app.
#### Screenshots (if appropriate)
image

#### Questions:
N/A

### CODE REVIEW

File: config/socket/game.js [hunk 1]
```diff
@@ -337,6 +337,7 @@ Game.prototype.pickCards = function(thisCardArray, thisPlayer) {
 Game.prototype.getPlayer = function(thisPlayer) {
   var playerIndex = this._findPlayerIndexBySocket(thisPlayer);
   if (playerIndex > -1) {
+    console.log(this.players[playerIndex]);
```
Comment (neddinn) [inline config/socket/game.js]: Take this off.

File: config/socket/socket.js [hunk 2]
```diff
@@ -104,6 +105,8 @@ module.exports = function(io) {
       player.avatar = avatars[Math.floor(Math.random()*4)+12];
       getGame(player,socket,data.room,data.createPrivate);
     }
+
+    console.log('join game called on socket class');
```
Comment (neddinn) [inline config/socket/socket.js]: This too

**Comment by fuzzybrows** [comment]
It looks good after stephen's comments

File: public/js/services/chat.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    /**
+    * Class to implement chat functionality
+    */
+    class Chat {
+      /**
+      * Constructor to create a new instance of this class
+      */
+      constructor() {
+        // declare fire base reference with link to our firebase database
+        this.firebase = new Firebase('https://anbu-cfh-chat.firebaseio.com/');
```
Comment (q-ode) [inline public/js/services/chat.js]: What do you think about moving this URL to and environment variable?

File: public/js/controllers/game.js [hunk 4]
```diff
@@ -8,6 +10,33 @@ angular.module('mean.system')
...
+    $scope.sendMessage = (userMessage) => {
+      $scope.chat.postGroupMessage(userMessage);
+      $scope.chatMessage = '';
+    };
+    /**
+    * Method to scroll the chat thread to the bottom
+    * so user can see latest message when messages overflow
+    * @return{undefined}
+    */
+    $scope.scrollChatThread = () => {
```
Comment (neddinn) [inline public/js/controllers/game.js]: You dont need to attach this to the scope. You can use var instead.

File: public/js/services/chat.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+          this.updateUnreadMessageCount();
+        });
+      }
+
+      /**
+      * Method to update the uread messages count
+      * @return{undefined} returns undefined
+      */
+      updateUnreadMessageCount() {
+        if (!this.showChatWindow) {
```
Comment (neddinn) [inline public/js/services/chat.js]: `showChatWindow` could be renamed to `isChatWindowShown`

File: public/js/controllers/game.js (hunk 4 above)
Comment (neddinn) [inline public/js/controllers/game.js]: You haven't addressed this yet. Please do not attach this to the scope.

File: public/js/controllers/game.js [hunk 6]
```diff
@@ -8,6 +10,33 @@ angular.module('mean.system')
...
+    $scope.sendMessage = (userMessage) => {
+      $scope.chat.postGroupMessage(userMessage);
+      $scope.chatMessage = '';
+    };
+    /**
+    * Method to scroll the chat thread to the bottom
+    * so user can see latest message when messages overflow
+    * @return{undefined}
+    */
+    scrollChatThread = () => {
```
Comment (neddinn) [inline public/js/controllers/game.js]: What houndci said :)

File: public/js/services/chat.js [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    /**
+    * Class to implement chat functionality
+    */
+    class Chat {
+      /**
+      * Constructor to create a new instance of this class
+      * @param{Object} dotEnv - Object holding our enviroment variables
+      * for angular
+      */
+      constructor(dotEnv) {
```
Comment (neddinn) [inline public/js/services/chat.js]: You can just call this `env` though, so we don't confuse it with the dotenv node module

File: public/js/constants.js [hunk 8]
```diff
@@ -0,0 +1,5 @@
+angular.module('mean.system', [])
+  .constant('env', {
+    FIREBASE_URL: 'https://anbu-cfh-chat.firebaseio.com/'
```
Comment (neddinn) [inline public/js/constants.js:3]: 👍

File: app/views/includes/foot.jade [hunk 9]
```diff
@@ -25,6 +25,7 @@ script(type='text/javascript', src='/lib/angular-ui-utils/modules/route.js')
 script(type='text/javascript', src='/js/init.js')
 script(type='text/javascript', src='/js/app.js')
 script(type='text/javascript', src='/js/directives.js')
+script(type='text/javascript', src='/js/constants.js')
 script(type='text/javascript', src='/js/filters.js')
 
 //firebase
```
Comment (andela-oakinseye) [inline app/views/includes/foot.jade:31]: Could  you remove the unneeded comment?

**Pull request merged**